components/provider_cldr/src/transform/dates.rs

```diff
@@ -132,6 +132,9 @@ impl From<&cldr_json::StylePatterns> for gregory::patterns::StylePatternsV1 {
 
 impl From<&cldr_json::DateTimeFormats> for gregory::patterns::DateTimeFormatsV1 {
     fn from(other: &cldr_json::DateTimeFormats) -> Self {
+        use gregory::patterns::{CountV1, SkeletonPatternV1, SkeletonsV1};
```

Comment:
I have a few thoughts here:

A: I was wondering about dependencies when I wrote it. I looked into using `PluralCategory` originally, but didn't understand the strategy for inter-dependencies between the various component crates. I could use the `PluralCategory` directly, but this would add `icu_plurals` as a dependency to `icu_datetime`. Is this OK?

B: I'm a bit concerned with sharing serializable representations with code outside of the provider. The `V1` signals a social contract to maintain the serialized represented. If I naively pulled in `PluralCategory`, then this gives no signal that it's used as a serializable source, and care should be taken when changing its representation. Perhaps this is fine here since this is a fairly standardized representation. However, is there still a risk of this assumption changing? This type of thing was a source of bugs when I maintained the Firefox Profiler.

C: I read through [data-pipeline.md](https://github.com/unicode-org/icu4x/blob/master/docs/design/data_pipeline.md) again, but I'm still not 100% sure on the guarantees here. Here in the provider, every struct has an associated version.

* How granular is this version number?
* Does this mean that individual pieces of the serialization can change? For example, what happens if I make a breaking change to `CountV1`, does that mean every struct that contains a `CountV1` needs to be bumped versions?
* Do we not care yet to define this breakdown, and label everything as V1 until something else is needed?
* Can we have something with a V2 key, but contain V1 structs?
* What happens when we have a breaking change?

Reply:
> A: I was wondering about dependencies when I wrote it. I looked into using PluralCategory originally, but didn't understand the strategy for inter-dependencies between the various component crates. I could use the PluralCategory directly, but this would add icu_plurals as a dependency to icu_datetime. Is this OK?

We could have it be an optional dependency and have skeletons be behind a feature (on by default, but this would allow you to build datetime with just styles for smaller footprint!)

Reply:
I mean, icu_datetime is going to have to depend on icu_plurals anyway in order to operate on this data, right?

In any case, don't have any problem with adding this dependency because icu_datetime is higher-level than icu_plurals.  It will also need to eventually gain a dependency on icu_numbers.

Reply:
On your other questions:

B: This would assume `PluralCategory` is serializable as an enum.  If it's not, then I would implement serde on that enum over in icu_plurals behind that crate's `serde` feature.  Then the stability guarantee comes from PluralCategory's serde form.

C: Let's not worry about data struct versioning until v1.  These are going to be interesting questions to discuss and answer when we actually have to start worrying about our stability guarantees.  Until that time, let's not worry about it much.

Reply:
> I mean, icu_datetime is going to have to depend on icu_plurals anyway in order to operate on this data, right?

What I'm saying is that `icu_datetime` that is only working with dateStyle/timeStyle is a usable date/time formatting API that may *not* pull in `icu_plurals`, am I correct?
`icu_plurals` become important when `skeletons` come into play.

Reply:
Just re-reading the style guide on crate features:

https://github.com/unicode-org/icu4x/blob/master/docs/process/style_guide.md#when-to-add-crate-features--suggested

> When adding enhancements to an ICU4X component, introduce features as a way for the end user to control the code size of their compilation as follows:
>
> 1. If the enhancement adds a new crate dependency, it should be behind a feature.
> 2. If the enhancement contains code that is not considered best practice, for example if it is mainly used for debugging diagnostics or development, then it should be behind a feature.

So it looks like this may fall under case 1.  My hope is that dead code elimination would take care of this one, but I have no problem with putting the skeleton handling stuff behind a feature flag.

Reply:
Given the conversation here, I'm doubting the use of storing the string data in memory at all. It would be fairly easy to implement `Serialize` and `Deserialize` for `Field` and `PatternItem`. Then the data could use something like the following shape:

```rs
pub type SkeletonTupleV1 = (
    SmallVec<[fields::Field; 5]>,
    SmallVec<[Vec<pattern::PatternItem>; 1]>,
);
```

Then all of the pattern matching machinery can take a lifetime parameter, and do all of its work via reference, without worrying about owning or copying any of this data. The data could be serialized back into the standard UTS 35 representations for strings, which is convenient and well-documented, but the in-memory representation would be exactly how we want to use it for the actual data processing.

The processing step can handle pre-sorting the information to make sure it's in the most efficient and direct form for skeleton matching. The skeleton matching machinery could then be a wrapper with a lifetime over this data representation.

Reply:
I am very much in favor of pre-processing CLDR strings such that the data stored in ICU4X DataProvider needs minimal parsing and processing at runtime.  I think it's okay if it incurs a small data size penalty as well.